voxblox/include/voxblox/utils/approx_hash_array.h

```diff
@@ -0,0 +1,153 @@
+#ifndef VOXBLOX_UTILS_APPROX_HASH_ARRAY_H_
+#define VOXBLOX_UTILS_APPROX_HASH_ARRAY_H_
+
+#include <atomic>
+#include <limits>
+#include "voxblox/core/common.h"
+
+// These classes allocate a fixed size array and index it with a hash that is
+// masked so that only its first N bits are non zero. This can be
+// thought of as a fast rough approximation of a hash table.
+// There are several advantages and some very significant disadvantages
+// Advantages-
+// - Simple and blazing fast lockless thread-safe approximate sets
+// - Can be used to provide more fine grain locking of blocks for threading then
+// simply locking the entire layer
+// Disadvantages-
+// - Highly inefficient use of memory (allocates 2^N elements)
+// - Cannot disern between two different elements with the same hash
+// - If the hash of two elements have the same first N elements of their hash,
+// only one can be store.
+
+// Basic container, given in an index and get the element that was stored there.
+// There are 2^unmasked_bits_ elements in the container, which element is
+// returned depends on your hashing function.
+// Uses at least 2^unmaksed_bits * sizeof(StoreElement) bytes of ram
+
+namespace voxblox {
+template <size_t unmasked_bits_, typename StoredElement>
+class ApproxHashArray {
```

Comment:
How confident are you that this is all correct? Because for such fundamental containers it would be nice to have at least some basic unit testing :)